File: DesignPatterns/2-strategyPattern/python/strategy.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# File: strategy.py
# Created Date: 2021-12-07 02:21:10
# Author: ysj
# Description: 策略类
###

from abc import ABCMeta, abstractmethod


class PriceCalculationStrategy(metaclass=ABCMeta):
    """抽象基类"""
    @abstractmethod
    def calculate_price(self, origin_price):
        pass


class KeepNormalStrategy(PriceCalculationStrategy):
    """原价策略"""

    def __init__(self):
        super().__init__()

    def calculate_price(self, origin_price):
        total_price = round(origin_price, 2)
        return total_price


class DiscountStrategy(PriceCalculationStrategy):
    """打折策略"""

    def __init__(self, discount):
        self.__discount = discount

    def calculate_price(self, origin_price):
        total_price = origin_price * self.__discount
        total_price = round(total_price, 2)
        return total_price


class FullReductionStrategy(PriceCalculationStrategy):
    """满减策略"""

    def __init__(self, full, reduction):
        self.__full = full
        self.__reduction = reduction

    def calculate_price(self, origin_price):
        if origin_price >= self.__full:
            origin_price -= self.__reduction
        total_price = round(origin_price, 2)
        return total_price


class DirectReductionStrategy(PriceCalculationStrategy):
    """无门槛策略"""

    def __init__(self, reduction):
        self.__reduction = reduction

    def calculate_price(self, origin_price):
        total_price = max(0, origin_price-self.__reduction)
        total_price = round(total_price, 2)
        return total_price

File: DesignPatterns/2-strategyPattern/typescript/strategy.ts
/**
 * -------------------------------------------------------
 * File: strategy.ts
 * Created Date: 2021-12-12 01:46:00
 * Author: ysj
 * Description: 策略类
 * -------------------------------------------------------
 */

/**策略接口 */
export interface IPriceCalculationStrategy {
  calculatePrice(originPrice: number): number;
}

/**原价策略 */
export class KeepNormalStrategy implements IPriceCalculationStrategy {
  calculatePrice(originPrice: number) {
    const totalPrice = Number(originPrice.toFixed(2));
    return totalPrice;
  }
}

/**打折策略 */
export class DiscountStrategy implements IPriceCalculationStrategy {
  private discount: number;
  constructor(discount: number) {
    this.discount = discount;
  }
  calculatePrice(originPrice: number) {
    let totalPrice = originPrice * this.discount;
    totalPrice = Number(totalPrice.toFixed(2));
    return totalPrice;
  }
}

/**满减策略 */
export class FullReductionStrategy implements IPriceCalculationStrategy {
  private full: number;
  private reduction: number;
  constructor(full: number, reduction: number) {
    this.full = full;
    this.reduction = reduction;
  }
  calculatePrice(originPrice: number) {
    let totalPrice =
      originPrice >= this.full ? originPrice - this.reduction : originPrice;
    totalPrice = Number(totalPrice.toFixed(2));
    return totalPrice;
  }
}

/**无门槛策略 */
export class DirectReductionStrategy implements IPriceCalculationStrategy {
  private reduction: number;
  constructor(reduction: number) {
    this.reduction = reduction;
  }
  calculatePrice(originPrice: number) {
    let totalPrice = Math.max(0, originPrice - this.reduction);
    totalPrice = Number(totalPrice.toFixed(2));
    return totalPrice;
  }
}
